app/controllers/application_controller.rb

```diff
@@ -14,17 +14,21 @@ class ApplicationController < ActionController::Base
 
   helper_method :blackcat_config
 
+  def login
+    fullpath = request.env.fetch('ORIGINAL_FULLPATH', '/')
+    session[:redirect_url] ||= fullpath == '/login' ? '/' : fullpath
+    session[:groups] = request.env.fetch(Settings.groups_header, '').split(',')
+    redirect_to current_user ? session[:redirect_url] : '/login'
```

Comment:
There's a lot going on in here that ought to be unpacked a bit. First off, I'd make a private method for the redirect url. And then avoid the other ternary:

``` ruby
def login
  session[:redirect_url] ||= home_or_original_path
  session[:groups] = request.env.fetch(Settings.groups_header, '').split(',')

  if current_user?
    redirect_to session[:redirect_url]
  else
    redirect_to login_path
  end
end
```

How does it not endless redirect back to itself if current_user is not present?

Reply:
I originally wrote with the if else, and anticipated it wasn't clever enough. :oof: 

If we run into a scenerio where there never becomes a current_user, the browser gets into a redirect loop. we could perhaps put a counter and after a threshold, redirect to a /401 -- that might be nicer to the user.

Reply:
There should always be a current user. It'll be a "guest user" from devise-guests.